### PR TITLE
Rocal performance PR

### PR DESCRIPTION
**Pull request by SundarRajan28**
- Added a graph attribute called num_threads which is set from master_graph which gets the value from the user during the pipeline creation
- Added single handle creation for all augmentations in amd_openvx_extensions
- Added a num_threads calculation which calculates the optimal threads to be used for a system in case the user doesn't provide the num_threads value
- Uses the num_threads value for openMP parallelization during decoding

### CODE REVIEW

File: amd_openvx/openvx/include/vx_ext_amd.h [hunk 1]
```diff
@@ -146,6 +146,8 @@ enum vx_graph_attribute_amd_e {
     VX_GRAPH_ATTRIBUTE_AMD_PERFORMANCE_INTERNAL_PROFILE = VX_ATTRIBUTE_BASE(VX_ID_AMD, VX_TYPE_GRAPH) + 0x07,
     /*! \brief OpenCL command queue. Use a <tt>\ref cl_command_queue</tt> parameter.*/
     VX_GRAPH_ATTRIBUTE_AMD_OPENCL_COMMAND_QUEUE         = VX_ATTRIBUTE_BASE(VX_ID_AMD, VX_TYPE_GRAPH) + 0x08,
+    /*! \brief OpenCL command queue. Use a <tt>\ref cl_command_queue</tt> parameter.*/
```
Comment (swetha097) [inline amd_openvx/openvx/include/vx_ext_amd.h]: Change the comment here - Its not OpenCL

File: rocAL/rocAL/include/api/rocal_api_data_transfer.h [hunk 2]
```diff
@@ -42,7 +42,7 @@ extern "C"  RocalStatus   ROCAL_API_CALL rocalCopyToOutputTensor(RocalContext ro
                                                               RocalTensorLayout tensor_format, RocalTensorOutputType tensor_output_type,
                                                               float multiplier0, float multiplier1, float multiplier2, float offset0,
                                                               float offset1, float offset2,
-                                                              bool reverse_channels);
+                                                              bool reverse_channels, bool normalization_on_device);
```
Comment (shobana-mcw) [inline rocAL/rocAL/include/api/rocal_api_data_transfer.h:45]: Can we rename this to normalization_on_GPU / GPU_normalization?
We used to set device as CPU / GPU. So this might be confusing.

File: rocAL/rocAL/include/pipeline/graph.h [hunk 3]
```diff
@@ -30,7 +30,7 @@ class Graph
 {
 public:
     enum class Status { OK = 0 };
-    Graph(vx_context context, RocalAffinity affinity, int cpu_id = 0, int gpu_id = 0 );
+    Graph(vx_context context, RocalAffinity affinity, int cpu_id = 0, size_t cpu_num_threads = 0, int gpu_id = 0 );
```
Comment (shobana-mcw) [inline rocAL/rocAL/include/pipeline/graph.h]: Set default value of CPU thread count to 1.

File: rocAL/rocAL/include/pipeline/master_graph.h [hunk 4]
```diff
@@ -94,8 +94,9 @@ class MasterGraph
         _output_images = output_images;
     }
     void set_output(Image* output_image);
+    void calculate_cpu_num_threads(size_t shard_count);
     bool empty() { return (remaining_count() < (_is_sequence_reader_output ? _sequence_batch_size : _user_batch_size)); }
-    size_t internal_batch_size() { return _internal_batch_size; }
+    size_t user_batch_size() { return _user_batch_size; }
```
Comment (shobana-mcw) [inline rocAL/rocAL/include/pipeline/master_graph.h]: Do we need an API for user_batch_size? I don't think so.
It's a private variable accessed by master graph. Check and remove if not needed.

File: rocAL/rocAL/include/pipeline/master_graph.h [hunk 5]
```diff
@@ -105,7 +106,6 @@ class MasterGraph
     bool is_sequence_reader_output() {return _is_sequence_reader_output; }
     void set_sequence_reader_output() { _is_sequence_reader_output = true; }
     void set_sequence_batch_size(size_t sequence_length) { _sequence_batch_size = _user_batch_size * sequence_length; }
-    void set_sequence_batch_ratio() { _sequence_batch_ratio = _sequence_batch_size / _internal_batch_size; }
     Status get_bbox_encoded_buffers(float **boxes_buf_ptr, int **labels_buf_ptr, size_t num_encoded_boxes);
```
Comment (shobana-mcw) [inline rocAL/rocAL/include/pipeline/master_graph.h:112]: Why sequence batch ratio is removed?

File: rocAL/rocAL/include/pipeline/master_graph.h [hunk 6]
```diff
@@ -94,8 +94,9 @@ class MasterGraph
         _output_images = output_images;
     }
     void set_output(Image* output_image);
+    void calculate_cpu_num_threads(size_t shard_count);
```
Comment (shobana-mcw) [inline rocAL/rocAL/include/pipeline/master_graph.h:97]: Does this function need public access?

File: rocAL/rocAL/include/pipeline/master_graph.h [hunk 7]
```diff
@@ -174,7 +172,6 @@ class MasterGraph
     std::vector<std::vector<size_t>> _sequence_start_framenum_vec; //!< Stores the starting frame number of the sequences.
     std::vector<std::vector<std::vector<float>>>_sequence_frame_timestamps_vec; //!< Stores the timestamps of the frames in a sequences.
     size_t _sequence_batch_size = 0; //!< Indicates the _user_batch_size when sequence reader outputs are required
-    size_t _sequence_batch_ratio; //!< Indicates the _user_to_internal_batch_ratio when sequence reader outputs are required
     bool _is_sequence_reader_output = false; //!< Set to true if Sequence Reader is invoked.
```
Comment (shobana-mcw) [inline rocAL/rocAL/include/pipeline/master_graph.h:179]: Why we are removing sequence batch ratio?
This is wrt video reader. Check and pls revert the changes.

File: rocAL/rocAL/source/api/rocal_api_meta_data.cpp [hunk 8]
```diff
@@ -231,8 +231,12 @@ ROCAL_API_CALL rocalGetImageLabels(RocalContext p_context, void* buf, unsigned i
     if(context->user_batch_size() != meta_data_batch_size)
         THROW("meta data batch size is wrong " + TOSTR(meta_data_batch_size) + " != "+ TOSTR(context->user_batch_size() ))
 
-    if (context->affinity == RocalAffinity::CPU)
-        memcpy(buf, meta_data.second->get_label_batch().data(), sizeof(int) * meta_data_batch_size);
+    if (context->affinity == RocalAffinity::CPU) {
+        // memcpy(buf, meta_data.second->get_label_batch().data(), sizeof(int) * meta_data_batch_size);
```
Comment (shobana-mcw) [inline rocAL/rocAL/source/api/rocal_api_meta_data.cpp]: We need to introduce a flag here too to copy to hip memory or else do a mem copy.

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 9]
```diff
@@ -940,21 +990,10 @@ MasterGraph::copy_output(unsigned char *out_ptr, size_t out_size_in_bytes)
 void MasterGraph::output_routine()
 {
     INFO("Output routine started with "+TOSTR(_remaining_count) + " to load");
```
Comment (shobana-mcw) [inline rocAL/rocAL/source/pipeline/master_graph.cpp:995]: Revert the Sequence reader changes.

**Review by shobana-mcw — COMMENTED** [review]
Pls resolve my review comments and let me know. 
We need to revert the sequence reader changes in few places.

File: rocAL/rocAL/include/api/rocal_api_data_transfer.h (hunk 2 above)
Comment (SundarRajan28) [reply rocAL/rocAL/include/api/rocal_api_data_transfer.h:45]: device is usually GPU while host is CPU. So that's the reason for the naming scheme. I'll change the variable name if its confusing.

File: rocAL/rocAL/include/pipeline/master_graph.h (hunk 4 above)
Comment (SundarRajan28) [reply rocAL/rocAL/include/pipeline/master_graph.h]: Its needed in rocal_api_data_loaders.cpp which previously called the internal_batch_size() function to get IBS value

File: rocAL/rocAL/include/pipeline/master_graph.h (hunk 5 above)
Comment (SundarRajan28) [reply rocAL/rocAL/include/pipeline/master_graph.h:112]: As discussed in the morning, this change is needed for some of the changes introduced in output_routine_video

File: rocAL/rocAL/include/pipeline/master_graph.h (hunk 6 above)
Comment (SundarRajan28) [reply rocAL/rocAL/include/pipeline/master_graph.h:97]: Its being called in rocal_api_dataloaders.cpp with shard count passed. Unless we make it public, calculating the num_threads from shard count variable is not possible during the loader creation unless we set shard count to the master_graph via a setter function and calculate it afterwards. But during one of our calls, you wanted me to set the num_threads as soon as the loader gets the shard count so this function is needed

File: rocAL/rocAL/include/pipeline/master_graph.h (hunk 7 above)
Comment (SundarRajan28) [reply rocAL/rocAL/include/pipeline/master_graph.h:179]: As discussed in the morning, this change is needed for some of the changes introduced in output_routine_video

File: rocAL/rocAL/source/pipeline/master_graph.cpp (hunk 9 above)
Comment (SundarRajan28) [reply rocAL/rocAL/source/pipeline/master_graph.cpp:995]: As discussed in the morning, this change is needed for some of the changes introduced in output_routine_video

File: rocAL/rocAL/include/pipeline/graph.h (hunk 3 above)
Comment (SundarRajan28) [reply rocAL/rocAL/include/pipeline/graph.h]: Resolved in the latest commit

File: amd_openvx/openvx/include/vx_ext_amd.h (hunk 1 above)
Comment (SundarRajan28) [reply amd_openvx/openvx/include/vx_ext_amd.h]: Resolved in the latest commit

File: rocAL/rocAL/source/api/rocal_api_meta_data.cpp (hunk 8 above)
Comment (SundarRajan28) [reply rocAL/rocAL/source/api/rocal_api_meta_data.cpp]: Resolved in the latest commit

File: rocAL/rocAL/include/pipeline/master_graph.h (hunk 6 above)
Comment (SundarRajan28) [reply rocAL/rocAL/include/pipeline/master_graph.h:97]: Removed this API and used context->user_batch_size() for the functions needing UBS value

File: amd_openvx_extensions/amd_rpp/source/nop.cpp [hunk 10]
```diff
@@ -102,24 +102,17 @@ static vx_status VX_CALLBACK initializeNop(vx_node node, const vx_reference *par
...
     STATUS_ERROR_CHECK(vxCopyScalar((vx_scalar)parameters[2], &data->device_type, VX_READ_ONLY, VX_MEMORY_TYPE_HOST));
 
-#if ENABLE_OPENCL
-    STATUS_ERROR_CHECK(vxQueryImage((vx_image)parameters[0], VX_IMAGE_ATTRIBUTE_AMD_OPENCL_BUFFER, &data->cl_pSrc, sizeof(data->cl_pSrc)));
-#elif ENABLE_HIP
-    STATUS_ERROR_CHECK(vxQueryImage((vx_image)parameters[0], VX_IMAGE_ATTRIBUTE_AMD_HIP_BUFFER, &data->hip_pSrc, sizeof(data->hip_pSrc)));
-#else
-    STATUS_ERROR_CHECK(vxQueryImage((vx_image)parameters[0], VX_IMAGE_ATTRIBUTE_AMD_HOST_BUFFER, &data->pSrc, sizeof(vx_uint8)));
-#endif
+    #if ENABLE_OPENCL
```
Comment (shobana-mcw) [inline amd_openvx_extensions/amd_rpp/source/nop.cpp]: Revert this change

File: amd_openvx_extensions/amd_rpp/source/copy.cpp [hunk 11]
```diff
@@ -25,7 +25,7 @@ THE SOFTWARE.
 struct CopyLocalData
 {
 
-    RPPCommonHandle handle;
+    RPPCommonHandle *handle;
```
Comment (shobana-mcw) [inline amd_openvx_extensions/amd_rpp/source/copy.cpp:28]: We can remove RPPCommonHandle if its not used. 
Same for nop node.

File: rocAL/rocAL/source/api/rocal_api_meta_data.cpp [hunk 12]
```diff
@@ -231,8 +231,16 @@ ROCAL_API_CALL rocalGetImageLabels(RocalContext p_context, void* buf, unsigned i
     if(context->user_batch_size() != meta_data_batch_size)
         THROW("meta data batch size is wrong " + TOSTR(meta_data_batch_size) + " != "+ TOSTR(context->user_batch_size() ))
 
-    if (context->affinity == RocalAffinity::CPU)
-        memcpy(buf, meta_data.second->get_label_batch().data(), sizeof(int) * meta_data_batch_size);
+    if (context->affinity == RocalAffinity::CPU) {
+        if (!flags) {
+            memcpy(buf, meta_data.second->get_label_batch().data(), sizeof(int) * meta_data_batch_size);
```
Comment (shobana-mcw) [inline rocAL/rocAL/source/api/rocal_api_meta_data.cpp:236]: Can we not directly assign it to pointer?

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 13]
```diff
@@ -1108,72 +1128,60 @@ void MasterGraph::output_routine_video()
...
-                        _output_images[idx]->swap_handle(write_buffers[idx]);
-                    else
-                    {
-                        auto this_cycle_buffer_ptr = (unsigned char *) write_buffers[idx] + each_cycle_size * cycle_idx;
-                        _output_images[idx]->swap_handle(this_cycle_buffer_ptr);
-                    }
-                }
+            // Swap handles on the output images, so that new processed image will be written to the a new buffer
+            for (size_t idx = 0; idx < _output_images.size(); idx++)
+            {
```
Comment (shobana-mcw) [inline rocAL/rocAL/source/pipeline/master_graph.cpp:1157]: Does this handle both Host and HIP buffer?

File: rocAL/rocAL/source/api/rocal_api_meta_data.cpp (hunk 12 above)
Comment (SundarRajan28) [reply rocAL/rocAL/source/api/rocal_api_meta_data.cpp:236]: Memcpy needs to happen otherwise the labels maybe overwritten during the next load routine while training is still running

File: rocAL/rocAL/source/pipeline/master_graph.cpp (hunk 13 above)
Comment (SundarRajan28) [reply rocAL/rocAL/source/pipeline/master_graph.cpp:1157]: Yes. I checked with Fiona and this is the way its handled in tensor pipeline

File: amd_openvx_extensions/amd_rpp/source/copy.cpp (hunk 11 above)
Comment (SundarRajan28) [reply amd_openvx_extensions/amd_rpp/source/copy.cpp:28]: Its needed for the single handle creation

File: amd_openvx_extensions/amd_rpp/source/nop.cpp (hunk 10 above)
Comment (SundarRajan28) [reply amd_openvx_extensions/amd_rpp/source/nop.cpp]: Reverted in the latest commit
